fixing issues with the cluster status computation

anvddriesch opened the pull request:
Towards: https://github.com/giantswarm/giantswarm/issues/15468

This adds more scenarios to the cluster status unit tests and fixes some errors in how the computation works.

## Checklist

- [ ] Update changelog in CHANGELOG.md.

anvddriesch commented on service/controller/resource/statuscondition/create.go:
```diff
@@ -286,22 +286,24 @@ func computeCreatedCondition(status infrastructurev1alpha2.CommonClusterStatus,
 	currentVersion := status.LatestVersion()
 	isCreated := status.HasCreatedCondition()
-	notUpdating := !status.HasUpdatingCondition()
+	notUpdating := status.LatestCondition() != infrastructurev1alpha2.ClusterStatusConditionUpdating
```
Here we used to ensure that the cluster was _never_ updating before. Now we ensure that the cluster is not _currently_ updating.

anvddriesch commented on service/controller/resource/statuscondition/create.go:
```diff
@@ -286,22 +286,24 @@ func computeCreatedCondition(status infrastructurev1alpha2.CommonClusterStatus,
-	isUpdating := status.HasUpdatingCondition()
-	notUpdated := !status.HasUpdatedCondition()
+	isUpdating := status.LatestCondition() == infrastructurev1alpha2.ClusterStatusConditionUpdating
+	notUpdated := status.LatestCondition() != infrastructurev1alpha2.ClusterStatusConditionUpdated
```
Same as above. We now check the _current_ status.

anvddriesch commented on service/controller/resource/statuscondition/create.go:
```diff
@@ -286,22 +286,24 @@ func computeCreatedCondition(status infrastructurev1alpha2.CommonClusterStatus,
+	isCreated := status.LatestCondition() == infrastructurev1alpha2.ClusterStatusConditionCreated
+	isUpdated := status.LatestCondition() == infrastructurev1alpha2.ClusterStatusConditionUpdated
+	hasTransitioned := isCreated || isUpdated
+	notSet := status.LatestVersion() != desiredVersion
```
We also check the _current_ version instead of all versions. If we roll back an update we may have the same version in the history twice.

anvddriesch commented on service/controller/resource/statuscondition/create.go:
```diff
@@ -286,22 +286,24 @@ func computeCreatedCondition(status infrastructurev1alpha2.CommonClusterStatus,
-	notSet := !status.HasVersion(desiredVersion)
+	isCreated := status.LatestCondition() == infrastructurev1alpha2.ClusterStatusConditionCreated
+	isUpdated := status.LatestCondition() == infrastructurev1alpha2.ClusterStatusConditionUpdated
+	hasTransitioned := isCreated || isUpdated
```
Again, we check the _current_ condition instead of all conditions.

anvddriesch commented on service/controller/resource/statuscondition/create_test.go:
```diff
@@ -6,30 +6,19 @@ import (
-	readyReplicas int
-}
-
 func TestComputeClusterStatusConditions(t *testing.T) {
```
I did not want to remove this test as it's still useful for including all the different computation checks. However, I simplified and removed the parts that are now not needed anymore.

anvddriesch commented on service/controller/resource/statuscondition/create_test.go:
```diff
@@ -550,14 +502,34 @@ func TestComputeUpdatingCondition(t *testing.T) {
 
 			expectedResult: false,
 		},
+		// we can also upgrade multiple times
```
This works now! Previously, an update would appear in the status only once

anvddriesch commented on service/controller/resource/statuscondition/create_test.go:
```diff
@@ -715,15 +703,57 @@ func TestComputeVersionChange(t *testing.T) {
 
 			expectedResult: false,
 		},
+		// the cluster has not transitioned yet
```
This works now, too! Previously, the version was already set before the update was done.

xh3b4sd commented on service/controller/resource/statuscondition/create_test.go:
```diff
@@ -715,15 +703,57 @@ func TestComputeVersionChange(t *testing.T) {
+					unittest.GetCreatedCondition(60),
+					unittest.GetCreatingCondition(90),
+				},
+				Versions: []infrastructurev1alpha2.CommonClusterStatusVersion{
+					unittest.GetVersion(20, "8.7.6"),
+					unittest.GetVersion(60, "8.7.5"),
+				},
```
Does this mean the scenario produces this version history? 

- 8.7.5
- 8.7.6
- 8.7.5

anvddriesch replied:
I expected that, as adding the new version is triggered here. However, it currently doesn't happen because in `apiextensions` the function `withVersion()` will only add the version to the history if it isn't already there. I am thinking to change that so we can reflect rollbacks in the history. WDYT?

xh3b4sd replied:
This should then be patched in `apiextensions` for sure. 👍

anvddriesch replied:
Yes, agreed

xh3b4sd reviewed (approved):
Super cool you fixed that.

njuettner reviewed (commented):
Good Job Antonia! Looks good to me if tested, basically I'm fine if we could run a test which triggers an update twice. IIRC if the cluster gets updated once it still worked in the past but broke when you upgraded again. So if we can see that this works on a test cluster, would be awesome :).

anvddriesch commented:
I have tested this quite extensively.
The conditions and versions are set correctly, even for several upgrades. This cluster was created and then upgraded twice:
```
    conditions:
    - condition: Updated
      lastTransitionTime: "2021-02-17T10:21:13Z"
    - condition: Updating
      lastTransitionTime: "2021-02-17T09:55:53Z"
    - condition: Updated
      lastTransitionTime: "2021-02-17T09:50:14Z"
    - condition: Updating
      lastTransitionTime: "2021-02-17T09:26:26Z"
    - condition: Created
      lastTransitionTime: "2021-02-17T09:24:28Z"
    - condition: Creating
      lastTransitionTime: "2021-02-17T09:07:32Z"
    id: vd7k8
    versions:
    - lastTransitionTime: "2021-02-17T10:21:13Z"
      version: 9.3.5
    - lastTransitionTime: "2021-02-17T09:50:14Z"
      version: 9.3.1-fix
    - lastTransitionTime: "2021-02-17T09:24:28Z"
      version: 9.3.1
```

The thing that still does not work is showing the version for a second time after the rollback. This cluster was upgraded once and then rolled back to the same version again:
```
    conditions:
    - condition: Updated
      lastTransitionTime: "2021-02-17T11:54:15Z"
    - condition: Updating
      lastTransitionTime: "2021-02-17T11:54:15Z"
    - condition: Updated
      lastTransitionTime: "2021-02-17T11:49:01Z"
    - condition: Updating
      lastTransitionTime: "2021-02-17T11:49:01Z"
    - condition: Updated
      lastTransitionTime: "2021-02-17T11:43:55Z"
    - condition: Updating
      lastTransitionTime: "2021-02-17T11:08:36Z"
    - condition: Updated
      lastTransitionTime: "2021-02-17T10:05:44Z"
    - condition: Updating
      lastTransitionTime: "2021-02-17T09:27:38Z"
    - condition: Created
      lastTransitionTime: "2021-02-17T09:19:15Z"
    - condition: Creating
      lastTransitionTime: "2021-02-17T09:08:28Z"
    id: 3yt4d
    versions:
    - lastTransitionTime: "2021-02-17T10:05:44Z"
      version: 9.3.1-fix
    - lastTransitionTime: "2021-02-17T09:19:15Z"
      version: 9.3.1
```
As you can see, the fact that the version is not set causes the updating and updated condition to appear repeatedly. 
This is something that has to be fixed in `apiextensions` and I will do so in a separate sub-PR.

xh3b4sd commented:
This is so cool to see. 😍
